src/com/facebook/buck/swift/SwiftDescriptions.java

```diff
@@ -16,60 +16,97 @@
-      TargetGraph targetGraph,
-      BuildRuleParams params,
-      BuildRuleResolver resolver,
-      CxxPlatform platform,
-      String moduleName,
-      ImmutableSortedSet<SourcePath> sources,
-      ImmutableList<String> compilerFlags,
-      ImmutableSortedSet<FrameworkPath> frameworks,
-      ImmutableSortedSet<FrameworkPath> libraries,
-      boolean enableObjcInterop) {
+      FlavorDomain<CxxPlatform> cxxPlatformFlavorDomain,
+      FlavorDomain<AppleCxxPlatform> platformFlavorsToAppleCxxPlatforms) {
+    BuildTarget parentTarget = parentParams.getBuildTarget();
+    CxxPlatform cxxPlatform = cxxPlatformFlavorDomain.getValue(parentTarget)
+        .or(defaultCxxPlatform);
+    // TODO(tho@uber.com) support version in swift flavor.
+    BuildTarget swiftCompanionTarget = parentTarget.withFlavors(
+        cxxPlatform.getFlavor(),
+        Optional.of("Swift").transform(Flavor.TO_FLAVOR).get());
+    BuildRuleParams params = parentParams.copyWithBuildTarget(swiftCompanionTarget);
+
+    // check the cache
+    Optional<BuildRule> rule = buildRuleResolver.getRuleOptional(swiftCompanionTarget);
+    if (rule.isPresent()) {
+      return rule;
+    }
+
+    if (!args.srcs.isPresent() || args.srcs.get().isEmpty()) {
+      return Optional.absent();
+    }
+
+    ImmutableSortedSet.Builder<SourcePath> swiftSrcsBuilder = ImmutableSortedSet.naturalOrder();
+    for (SourceWithFlags source : args.srcs.get()) {
+      if (MoreStrings.endsWithIgnoreCase(source.getSourcePath().toString(), SWIFT_EXTENSION)) {
+        swiftSrcsBuilder.add(source.getSourcePath());
+      }
+    }
+    ImmutableSortedSet<SourcePath> swiftSrcs = swiftSrcsBuilder.build();
+    if (swiftSrcs.isEmpty()) {
+      return Optional.absent();
+    }
+
     AppleCxxPlatform appleCxxPlatform = ApplePlatforms.getAppleCxxPlatformForBuildTarget(
         cxxPlatformFlavorDomain,
         defaultCxxPlatform,
-        appleCxxPlatformFlavorDomain,
+        platformFlavorsToAppleCxxPlatforms,
         params.getBuildTarget(),
         Optional.<MultiarchFileInfo>absent());
     Optional<Tool> swiftCompiler = appleCxxPlatform.getSwift();
     if (!swiftCompiler.isPresent()) {
       throw new HumanReadableException("Platform %s is missing swift compiler", appleCxxPlatform);
     }
-    return new SwiftCompile(
-        params,
-        new SourcePathResolver(resolver),
+
+    return Optional.<BuildRule>of(new SwiftLibrary(
         swiftCompiler.get(),
-        moduleName,
-        BuildTargets.getGenPath(params.getProjectFilesystem(), params.getBuildTarget(), "%s"),
-        sources);
+        params,
+        new SourcePathResolver(buildRuleResolver),
+        ImmutableList.<BuildRule>of(),
+        args.frameworks.get(),
+        args.libraries.get(),
+        platformFlavorsToAppleCxxPlatforms,
+        BuildTargets.getGenPath(
+            params.getProjectFilesystem(),
+            swiftCompanionTarget.withAppendedFlavors(cxxPlatform.getFlavor()), "%s"),
```

Comment:
Don't need `.withAppendedFlavors` as you already have the platform flavor in `swiftCompanionTarget`.